src/components/demos/write/Demo.js

```diff
@@ -21,138 +24,122 @@ const Wrapper = styled.div`
 const failedLogin = (apiOutput) =>
   extractLoginResult(apiOutput) === 'Login failed'
 const loginError = (apiOutput) => extractLoginResult(apiOutput) === 'error'
 
-const SecretDesktop = ({ icons }) => {
-  return (
-    <Col styles={{ padding: '0px 10px', alignItems: 'flex-start' }}>
-      {icons.map((icon) => (
-        <IconWrapper key={icon.name}>
-          <img
-            src={icon.img}
-            style={{ cursor: 'pointer' }}
-            width={'60px'}
-            height={'60px'}
-          />
-          <IconText>{icon.name}</IconText>
-        </IconWrapper>
-      ))}
-    </Col>
-  )
-}
-
 export default function WriteDemo(props) {
   const { execute, binaryName, helpContent } = props
 
   const nav = useNavigate()
   const state = React.useContext(Context)
-  const { update, writeDemo: demoState } = state
-  const { theme } = demoState
+  const { update, writeDemo } = state
+  const { theme, showHelp, usernamePasswordPairs, output, showResponse } =
+    writeDemo
+  const demoState = { ...props, ...writeDemo }
   const isMorello = theme.name === 'Morello'
-
-  const [animateLoginFailed, setAnimateLoginFailed] = useState(false)
-  const [showHelp, setShowHelp] = useState(false)
-  const [awaitingApi, setAwaitingApi] = useState(false)
-  const [apiOutput, setApiOutput] = useState('')
-  const [usernamePasswordPairs, setUsernamePasswordPairs] = useState([])
-
-  const resetStates = () => {
-    setShowHelp(false)
-    setAwaitingApi(false)
-    setApiOutput('')
-    setUsernamePasswordPairs([])
-  }
+  const bin = `${binaryName}-${theme.arch}`
+  const [animateLoginFailed, setAnimateLoginFailed] = React.useState(false)
 
   const switchToMorello = (e) => {
     e.preventDefault()
-    resetStates()
     update({
       writeDemo: {
         ...demoState,
+        showHelp: false,
+        fetching: false,
+        output: undefined,
+        showResponse: false,
+        usernamePasswordPairs: [],
         theme: Themes('Morello'),
       },
     })
   }
 
-  useEffect(() => {
-    update(initState)
-  }, [update])
-
   useEffect(() => {
     const attemptLogin = async () => {
-      setAwaitingApi(true)
-      const output = await execute(
-        `${binaryName}-${theme.arch}`,
-        usernamePasswordPairs
-      )
-      setApiOutput(output)
-      setAwaitingApi(false)
+      update({
+        writeDemo: {
+          ...demoState,
+          fetching: true,
+        },
+      })
+      // TODO allow execute to handle states e.g. isFetching isError...
+      const output = await execute(bin, usernamePasswordPairs)
+      update({
+        writeDemo: {
+          ...demoState,
+          fetching: false,
+          output,
+        },
+      })
     }
 
     if (usernamePasswordPairs.length > 0) {
       attemptLogin()
     }
-  }, [usernamePasswordPairs, execute, binaryName, theme])
+    // TODO address demoState dependency, gets trapped in re-rendering state
+  }, [bin, execute, update, usernamePasswordPairs])
```

Comment:
linting error, missing `demoState` in deps

Reply:
yea, bus this is why I left a. comment still there, to be fair we should have a method or ctx for each demo.

Reply:
@mattdean-digicatapult this is the only one I would like to leave as it's for time being, there is a comment as well. If you got any ideas - please do let me know.

I would be happy to break down ctx into per demo basis so less spreading would be required

Reply:
Yeah this concerns me. It basically means prop handling is broken and leading to re-rendering

Reply:
true, let me see if I can fix this relatively simply. I can get it working by adding another 10 lines, basically deconstructing object to what we need only and not using whole object. This way it does not get into that re-render state.

However, I believe that better solution would be to give each demo it's dedicated context or create a different update method, so far we have just one ...

Reply:
this has been resolved by adding isFetching to the root of ctx